### PR TITLE
[WIP] Update tutorials

### PR DESCRIPTION
**Pull request by AlexDenisov**
(I didn't do any proof-reading, just dumped the thing)

### CODE REVIEW

File: docs/HelloWorld.rst [hunk 1]
```diff
@@ -238,67 +252,77 @@ The code:
...
+          bool test2 = valid_age(20) == false;
+                                     ^
+    /tmp/sc-PzmaCNIRu/main.cpp:21:34: warning: Killed: Replaced == with != [cxx_eq_to_ne]
+          bool test3 = valid_age(21) == true;
+                                     ^
     [info] All mutations have been killed
     [info] Mutation score: 100%
+    [info] Total execution time: 554ms
+
+In this last run we see that all mutants were killed since we covered with tests
```
Comment (stanislaw) [inline docs/HelloWorld.rst:314]: In this last run,

File: docs/HelloWorld.rst [hunk 2]
```diff
@@ -238,67 +252,77 @@ The code:
...
+all cases around the ``<=``.
 
 Summary
 -------
 
-This is a short summary of what we have learned in tutorial:
-
-- Your code has to be compiled with ``-fembed-bitcode -g`` compile flags:
+This is a short summary of what we have learned in tutorial.
```
Comment (stanislaw) [inline docs/HelloWorld.rst:320]: the tutorial 😄

File: docs/HelloWorld.rst [hunk 3]
```diff
@@ -238,67 +252,77 @@ The code:
...
+Your code has to be compiled with ``-fembed-bitcode -g`` compile flags:
 
   - Mull expects embedded bitcode files to be present in a binary executable
     (ensured by ``-fembed-bitcode``).
 
   - Mull needs debug information to be included by the compiler (enabled by
     ``-g``). Mull uses this information to find mutations in bitcode and source
     code.
-
-- Mull expects the following arguments to be always provided:
```
Comment (stanislaw) [inline docs/HelloWorld.rst:302]: Why did you remove this one?

File: docs/SQLiteReport.rst [hunk 4]
```diff
@@ -1,6 +1,8 @@
 Working with SQLite report
 ==========================
 
+**Warning: the data model has changed and most of the tutorial is not longer relevant/applicable**
```
Comment (stanislaw) [inline docs/SQLiteReport.rst:4]: no longer

File: docs/Tutorials.rst [hunk 5]
```diff
@@ -6,5 +6,5 @@ Tutorials
 
    HelloWorld
    fmtlibTutorial
-   OpenSSLTutorial
+   controlMutationsTutorial
```
Comment (stanislaw) [inline docs/Tutorials.rst:9]: Control? (capitalize)

File: docs/controlMutationsTutorial.rst [hunk 6]
```diff
@@ -0,0 +1,207 @@
+Keeping mutants under control
+=============================
+
+This tutorial will show you how to keep the amount of mutants under control.
+It builds on top of the `fmtlib tutorial <fmtlibTutorial.html>`_ so make sure
+you go through it first.
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:6]: through that one?

File: docs/controlMutationsTutorial.rst [hunk 7]
```diff
@@ -0,0 +1,207 @@
+Keeping mutants under control
+=============================
+
+This tutorial will show you how to keep the amount of mutants under control.
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:4]: tutorial shows how

File: docs/controlMutationsTutorial.rst [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+It builds on top of the `fmtlib tutorial <fmtlibTutorial.html>`_ so make sure
+you go through it first.
+
+When you apply mutation testing for the first time you might be overwhelmed by
+the amount of mutants - what do you do when you see that several hundreds or
+thousands of mutants survived?
+
+The right way to go about it is to put the number of mutants under control and
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:12]: Maybe "one way"?

File: docs/controlMutationsTutorial.rst [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+work through them incrementally.
+
+Mutation Operators
+------------------
+
+If you apply Mull with the default set of mutation operators on fmtlib you will get
+around ~4000 mutants, ~3300 of which survived.
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:19]: survive

File: docs/controlMutationsTutorial.rst [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+                                                      ^
+    [info] Mutation score: 13%
+    [info] Total execution time: 89999ms
+
+Obviously, going through all of them to see which ones deserve your attention is
+simply impractical.
+
+The easiest way to decrease this number is to pick one or two `mutation operators <SupportedMutations.html>`_.
+
+Let's see how the numbers change if we only use ``cxx_add_to_sub`` which replaces
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:55]: that replaces

File: docs/controlMutationsTutorial.rst [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+
+    /tmp/sc-0Puh0WBoL/fmt/test/gmock-gtest-all.cc:1758:34: warning: Survived: Replaced + with - [cxx_add_to_sub]
+      state_ = (1103515245U * state_ + 12345U) % kMaxRange;
+                                     ^
+    /tmp/sc-0Puh0WBoL/fmt/test/gmock-gtest-all.cc:2275:55: warning: Survived: Replaced + with - [cxx_add_to_sub]
+      return static_cast<TimeInMillis>(now.tv_sec) * 1000 + now.tv_usec / 1000;
+                                                          ^
+
+Looking at the paths it is clear that these mutants are part of the GoogleTest
+framework (``gmock-gtest-all.cc``). It is very unlikely that you are not interested
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:108]: that you are interested

File: docs/controlMutationsTutorial.rst [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+in seeing this in the result.
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:109]: these?

File: docs/controlMutationsTutorial.rst [hunk 13]
```diff
@@ -0,0 +1,207 @@
...
+Mull comes with two path-based filters ``--exclude-path`` and ``--include-path``.
+You can use these to either exclude or include mutations based on their file-system location.
+Let's exclude anything related to GoogleTest:
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:112]: anything -> everything

File: docs/controlMutationsTutorial.rst [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+      auto it = reserve(out, size + padding * specs.fill.size());
+                                  ^
+    [info] Mutation score: 48%
+    [info] Total execution time: 14124ms
+
+
+Now, we've got only 27 mutants instead of 4000 in the beginning - something we can
+work with.
+In fact, it's always a good idea to start with the code coverage in the first place.
+In this case, even without filters and changing the set of mutation operators we
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:201]: operators,

**Review by stanislaw — CHANGES_REQUESTED** [review]

File: docs/controlMutationsTutorial.rst [hunk 15]
```diff
@@ -0,0 +1,207 @@
+Keeping mutants under control
+=============================
+
+This tutorial will show you how to keep the amount of mutants under control.
+It builds on top of the `fmtlib tutorial <fmtlibTutorial.html>`_ so make sure
+you go through it first.
+
+When you apply mutation testing for the first time you might be overwhelmed by
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:8]: the first time,

File: docs/controlMutationsTutorial.rst [hunk 16]
```diff
@@ -0,0 +1,207 @@
...
+the amount of mutants - what do you do when you see that several hundreds or
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:9]: checked with Grammarly just this one: the number of mutants - it recommends 😄

File: docs/controlMutationsTutorial.rst (hunk 9 above)
Comment (stanislaw) [reply docs/controlMutationsTutorial.rst:19]: on fmtlib,

File: docs/controlMutationsTutorial.rst [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+    /tmp/sc-0Puh0WBoL/fmt/include/fmt/format-inl.h:1334:68: warning: Survived: Replaced + with - [cxx_add_to_sub]
+      int num_bigits() const { return static_cast<int>(bigits_.size()) + exp_; }
+                                                                       ^
+    /tmp/sc-0Puh0WBoL/fmt/include/fmt/format-inl.h:1284:53: warning: Survived: Replaced + with - [cxx_add_to_sub]
+          double_bigit result = bigits_[i] * wide_value + carry;
+                                                        ^
+    [info] Mutation score: 12%
+    [info] Total execution time: 18481ms
+
+You still get plenty - 305 survived out of 350 total, but this is much more
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:90]: You are still getting plenty?

File: docs/controlMutationsTutorial.rst [hunk 18]
```diff
@@ -0,0 +1,207 @@
...
+275/320 vs 305/350. Better, but still too much.
+
+Code Coverage
+-------------
+
+In fact, many of the survived mutants can never be detected by the test suite because
+they are not reachable by any of the tests. We can leverage code coverage information
+to cut off all those mutants.
+
+For that to work we need to gather the coverage info first.
```
Comment (stanislaw) [inline docs/controlMutationsTutorial.rst:145]: to work,

File: docs/HelloWorld.rst (hunk 2 above)
Comment (AlexDenisov) [reply docs/HelloWorld.rst:320]: articles overrated 😄

File: docs/HelloWorld.rst (hunk 3 above)
Comment (AlexDenisov) [reply docs/HelloWorld.rst:302]: There were two of them before, now there is only one so I find it weird to have this paragraph here. Shall I put it back?

File: docs/Tutorials.rst (hunk 5 above)
Comment (AlexDenisov) [reply docs/Tutorials.rst:9]: yeah, just some odd typo.

File: docs/controlMutationsTutorial.rst (hunk 8 above)
Comment (AlexDenisov) [reply docs/controlMutationsTutorial.rst:12]: What is the other way? :)

File: docs/HelloWorld.rst (hunk 3 above)
Comment (stanislaw) [reply docs/HelloWorld.rst:302]: I understand now. I don't think it is weird, it just says the truth - you only need to provide an executable. Use your best judgement 👍
